NWNXLib/Services/Commands/Commands.cpp

```diff
@@ -0,0 +1,144 @@
+#include "Services/Commands/Commands.hpp"
+#include <stdio.h>
+#include <dlfcn.h>
+
+namespace NWNXLib {
+namespace Services {
+
+static Commands *g_commands;
+
+Commands::Commands()
+{
+    g_commands = this;
+    RegisterCommand("nwnx_commands_test", [](std::string& s){ LOG_NOTICE("Selftest command. Args: '%s'", s.c_str()); });
+}
+
+Commands::~Commands()
+{
+    g_commands = nullptr;
+
+}
+
+bool Commands::RegisterCommand(const std::string& cmd, CommandFunc func)
+{
+    auto it = m_commandMap.find(cmd);
+    if (it == m_commandMap.end())
+    {
+        m_commandMap[cmd] = func;
+        LOG_INFO("Registering command '%s'", cmd.c_str());
+        return true;
+    }
+    else
+    {
+        LOG_WARNING("Command '%s' already registered", cmd.c_str());
+        return false;
+    }
+}
+
+void Commands::UnregisterCommand(const std::string& cmd)
+{
+    LOG_INFO("Unregistering command '%s'", cmd.c_str());
+    m_commandMap.erase(cmd);
+}
+
+// This runs on a side thread that processes input.
+bool Commands::ScheduleCommand(std::string cmdline)
+{
+    auto trim = [](std::string &s)
+    {
+        s.erase(0, s.find_first_not_of(" \n\r\t"));
+        s.erase(s.find_last_not_of(" \n\r\t") + 1);
+    };
+
+    trim(cmdline);
+    std::string cmd, args;
+    size_t space = cmdline.find(' ');
+
+    cmd = cmdline.substr(0, space);
+    args = (space == std::string::npos) ? "" : cmdline.substr(space, std::string::npos);
+
+    trim(cmd); trim(args);
+
+    auto it = m_commandMap.find(cmd);
+    if (it != m_commandMap.end())
+    {
+        std::lock_guard<std::mutex> guard(m_queueLock);
+        m_commandQueue.emplace_back(cmd, args);
+        LOG_DEBUG("Scheduled command '%s' with args '%s'", cmd.c_str(), args.c_str());
+        return true;
+    }
+
+    return false;
+}
+
+void Commands::RunScheduledCommands()
+{
+    std::lock_guard<std::mutex> guard(m_queueLock);
+    for (auto item: m_commandQueue)
```

Comment:
You could `std::swap` `m_commandQueue` with a vector on the stack and then release the lock early to prevent the scheduling thread from blocking if the command is expensive. It probably doesn't matter much, I just thought I'd mention it because I mentioned the other thing too.

Reply:
Eh, I don't really want to. I don't care if we block the input thread for however long, it isn't doing anything useful anyway - can't collect new input while the current command is being processed. And adding the swap() will cause extra work on the main thread, which is on the critical path.

Reply:
Oh, I had it the wrong way around. For some reason I thought we were blocking the main thread.